crates/red_knot_python_semantic/src/types/class.rs

```diff
@@ -792,21 +792,29 @@ impl<'db> ClassLiteralType<'db> {
     /// traverse through the MRO until it finds the member.
     pub(super) fn own_class_member(self, db: &'db dyn Db, name: &str) -> SymbolAndQualifiers<'db> {
         if let Some(metadata) = self.dataclass_metadata(db) {
-            if name == "__init__" {
-                if metadata.contains(DataclassMetadata::INIT) {
-                    // TODO: Generate the signature from the attributes on the class
-                    let init_signature = Signature::new(
-                        Parameters::new([
-                            Parameter::variadic(Name::new_static("args"))
-                                .with_annotated_type(Type::any()),
-                            Parameter::keyword_variadic(Name::new_static("kwargs"))
-                                .with_annotated_type(Type::any()),
-                        ]),
-                        Some(Type::none(db)),
+            if name == "__init__" && metadata.contains(DataclassMetadata::INIT) {
+                // TODO: Generate the signature from the attributes on the class
+                let init_signature = Signature::new(
+                    Parameters::new([
+                        Parameter::variadic(Name::new_static("args"))
+                            .with_annotated_type(Type::any()),
+                        Parameter::keyword_variadic(Name::new_static("kwargs"))
+                            .with_annotated_type(Type::any()),
+                    ]),
+                    Some(Type::none(db)),
+                );
+
+                return Symbol::bound(Type::Callable(CallableType::new(db, init_signature))).into();
+            } else if matches!(name, "__lt__" | "__le__" | "__gt__" | "__ge__") {
+                if metadata.contains(DataclassMetadata::ORDER) {
+                    let signature = Signature::new(
+                        Parameters::new([Parameter::positional_or_keyword(Name::new_static(
+                            "other",
+                        ))
+                        .with_annotated_type(Type::instance(self.default_specialization(db)))]),
+                        Some(KnownClass::Bool.to_instance(db)),
                     );
-
-                    return Symbol::bound(Type::Callable(CallableType::new(db, init_signature)))
-                        .into();
+                    return Symbol::bound(Type::Callable(CallableType::new(db, signature))).into();
```

Comment:
@dcreager Here, I'm synthesizing a signature for autogenerated `__lt__` methods (and similar) for dataclasses. I think the `other` argument should have an annotated type of `Self`, such that instances of the dataclass can be compared with `other` instances, but not with completely unrelated types. Since we don't have `Self` yet, I tried to construct the `Type::Instance(…)` type that corresponds to this `ClassLiteralType`. Would you mind taking a look at the `self.default_specialization(db)` call here to see if that looks correct?

Reply:
> And you might want to add a new `Type::apply_optional_specialization` to make that a bit more ergonomic

I wasn't sure what you were suggesting here. Which part could be made more ergonomic? The `Type::instance(self.apply_optional_specialization(db, specialization))` construction?